crates/codegen/parser/runtime/src/query/engine.rs

```diff
@@ -0,0 +1,481 @@
+use std::collections::HashMap;
+use std::rc::Rc;
+
+// This crate is copied to another crate, so all imports should be relative
+use super::super::cst;
+use super::super::cursor::Cursor;
+use super::model::{
+    AlternativesMatcher, BindingMatcher, Kind, Matcher, NodeMatcher, NodeSelector,
+    OneOrMoreMatcher, OptionalMatcher, Query, SequenceMatcher,
+};
+
+impl Cursor {
+    pub fn query(self, queries: Vec<Query>) -> QueryResultIterator {
+        QueryResultIterator::new(self, queries)
+    }
+
+    fn consume(&mut self) -> bool {
```

Comment:
not sure I understand the difference between `Cursor::consume()` and `Cursor::complete()` from the names. The first only completes the current node, right? it won't actually consume the cursor, and it doesn't have to be completed afterwards.

Reply:
Renamed to make it more obvious. The tricky bit is mixing the return value and the action. This is specific to the query execution semantics. I put it here for DRY.